### ISSUE TITLE
Bad performance on lot of single updates

### ISSUE THREAD

**Issue by Tiller** [spring-projects/spring-data-mongodb#3760]
Hello,

While I was rewriting an existing project using reactor and spring-data mongodb, I noticed a big performance decrease which led me to create some performance tests.

You can find the said tests and a result sample at:
https://github.com/Tiller/mongo-perfs

But summary is:
| Test                                                                               | Mean Time (s) | Diff Base Driver | Diff Base Reactive Driver |
|------------------------------------------------------------------------------------|---------------|------------------|---------------------------|
| Java Driver ([testRaw](https://github.com/Tiller/mongo-perfs/blob/master/src/test/java/com/sigfox/monitoring/MongoPerfTest.java#L70))                                                              | 3.320         | 🟢 +0%            | 🟢 -28%                    |
| Java Reactive Driver ([testReactiveRaw](https://github.com/Tiller/mongo-perfs/blob/master/src/test/java/com/sigfox/monitoring/MongoPerfTest.java#L168))                                             | 4.627         | 🔴 +39%           | 🟢 +0%                     |
| Reactor Single Block ([testBlockExecutor](https://github.com/Tiller/mongo-perfs/blob/master/src/test/java/com/sigfox/monitoring/MongoPerfTest.java#L157))                                           | 4.793         | 🔴 +44%           | 🔴 +4%                     |
| Reactor Flat Map w/ Concurrency ([testFlatMapWithConcurrency](https://github.com/Tiller/mongo-perfs/blob/master/src/test/java/com/sigfox/monitoring/MongoPerfTest.java#L115))                       | 8.858         | 🔴 +167%          | 🔴 +91%                    |
| Reactor Flat Map w/ Big Concurrency ([testFlatMapWithBigConcurrency](https://github.com/Tiller/mongo-perfs/blob/master/src/test/java/com/sigfox/monitoring/MongoPerfTest.java#L135))                | 8.882         | 🔴 +168%          | 🔴 +92%                    |
| Reactor Flat Map w/ Concurrency & Prefetch ([testFlatMapWithConcurrencyAndPrefetch](https://github.com/Tiller/mongo-perfs/blob/master/src/test/java/com/sigfox/monitoring/MongoPerfTest.java#L125)) | 9.024         | 🔴 +172%          | 🔴 +95%                    |
| Reactor Parallel ([testParallel](https://github.com/Tiller/mongo-perfs/blob/master/src/test/java/com/sigfox/monitoring/MongoPerfTest.java#L145))                                                    | 8.970         | 🔴 +170%          | 🔴 +94%                    |

Note: this result is over 5 iterations of each test, removing the first iteration. I ran these 5 iterations multiple times today, and I've always had similar results

Basically I'm doing 300'000 single updates over 200 collections using 12 threads/workers, and what I'm observing is (aside the +39% increase in time comparing the java driver to the java reactive driver) a +90% increase in the processing time when using the spring-data reactor layer and flux.

However, as I only have a 4% increase when using `.block()` on each updates, I wonder whether this is a spring-data issue, or a reactor issue (or obviously a mistake I made in my testing).

Have you observed such results when doing performance tests? Am I doing something wrong? Can I improve on something?

Thanks

Note: I know using `@RepeatedTest` is not necessarily the best way of writing perf tests, but I don't think it change the result that much anyway. And yes, I'm using the default client configuration, and my DB is local. But I observed the same performance decrease on a production environment with dedicated shards

**Comment by Julien-Eyraud** [comment]
I created a micro benchmark with JMH here https://github.com/Julien-Eyraud/spring-data-mongo-benchmark in order to help on this issue.
Here are the result for 2 warmup iterations and 5 iterations on a AWS t3.2xlarge (mongodb 4.0.19 running inside a container on the same instance) : 
```bash
docker run -d -p 27017:27017 mongo:4.0.19
java -jar benchmarks.jar  -wi 2 -i 5 -bm avgt
```
Benchmark | Avg (sec/ops) | Diff Base Driver
------------ | -------------|-------------
[syncRaw](https://github.com/Julien-Eyraud/spring-data-mongo-benchmark/blob/d8ceff0e23010033a2ad3ade34c1b6cbcbe7b516/src/main/java/com/sigfox/oss/MyBenchmark.java#L148)| 6.921 ± 0.349| +0% :white_circle: 
[mongoTemplate](https://github.com/Julien-Eyraud/spring-data-mongo-benchmark/blob/d8ceff0e23010033a2ad3ade34c1b6cbcbe7b516/src/main/java/com/sigfox/oss/MyBenchmark.java#L105)| 9.017 ± 0.947 | +30% :red_circle: 
[reactiveRaw](https://github.com/Julien-Eyraud/spring-data-mongo-benchmark/blob/d8ceff0e23010033a2ad3ade34c1b6cbcbe7b516/src/main/java/com/sigfox/oss/MyBenchmark.java#L89)| 11.417 ± 0.381 | +64.9% :red_circle: 
[reactiveTemplate](https://github.com/Julien-Eyraud/spring-data-mongo-benchmark/blob/d8ceff0e23010033a2ad3ade34c1b6cbcbe7b516/src/main/java/com/sigfox/oss/MyBenchmark.java#L80) | 20.782 ± 0.429 | +200% :red_circle:

**Comment by mp911de** [comment]
Thanks a lot for the report and the benchmarks. I had a look at the benchmarks and ran these against a local MongoDB instance to avoid measurement errors caused by AWS noisy neighbors or network issues. 
I found a few issues with the benchmarks themselves:

1. Heavy usage of executors/schedulers: Using `ExecutorService` or switching threads using `subscribeOn(…)`/`publishOn(…)` adds unnecessary load to the benchmark that can result in jitter and it isn't essential to the MongoDB driver
2. Improper batching: `NB_UPDATE` accumulates updates through a Java `Stream`. Streams themselves impose quite some GC and CPU pressure by their sole usage. Benchmarks should ideally use un-batched scenarios and be limited to the benchmark subject to avoid impact from components that are not part of the system under benchmark.
3. Debug logging: Log4j is configured with `DEBUG` log levels. The MongoDB driver doesn't make much use of logging but Spring Data does (e.g. logging the queries being ran against MongoDB)


By fixing these issues I end up with a result of:

```
Benchmark                      Mode  Cnt      Score      Error  Units
MyBenchmark.mongoTemplate     thrpt    3  11963,897 ± 1171,901  ops/s
MyBenchmark.reactiveRaw       thrpt    3  11338,516 ± 7639,737  ops/s
MyBenchmark.reactiveTemplate  thrpt    3   8450,458 ± 7055,813  ops/s
MyBenchmark.syncRaw           thrpt    3  13389,916 ± 7742,427  ops/s
```
 
Benchmark | Avg (ops/sec) | Diff Base Driver| Diff Base Reactive Driver
------------ | -------------|-------------|-------------
syncRaw| 13389,916 ± 7742,427| +0% | n/a
mongoTemplate| 11963,897 ± 1171,901 | +11%| n/a
reactiveRaw|  11338,516 ± 7639,737 | +18% | +0%
reactiveTemplate| 8450,458 ± 7055,813 | +58% | +34%


The relations between those components are much smaller, however, the overhead of `MongoTemplate` compared to its baseline (synchronous driver) is about 11% whereas the overhead of `ReactiveMongoTemplate` compared to its baseline (reactive driver) is about 34%. We need to investigate where the majority of time is spent and check whether we can improve.


You can find the updated benchmark at https://gist.github.com/mp911de/82df815fadf5b7646ba7474d598c75c0.

**Comment by mp911de** [comment]
I found a few items that we could improve, mostly bypassing transaction participation that speeds up the use of `ReactiveMongoTemplate` by another 10%. These are going to be addressed with #3809.
